### PR TITLE
Introduce the ability for themes to provide custom icons for the Icon block

### LINKED ISSUE #923: ISBAT add custom icons or choose from larger selection of Icons
**Is your feature request related to a problem? Please describe.**
The Icon block has great functionalities with the ability to resize and set custom colors. The block could be improved if a user had an easy way to add custom icons. If not adding custom icons, perhaps we should explore expanding the selection of available Scalable Vector Graphics to use as icons.

**Describe alternatives you've considered**
Alternatives to additional icons would be to use a regular image with Image block or to use SVG markup within an HTML block. The negative with this alternative is that the user cannot set custom colors easily and cannot adjust the size in a scalable way.

**Additional context**
A few ideas for additional icons.
- Sports (championship, teams, people etc.)
- Brands (Google, Facebook, Amazon, Microsoft etc.)
- Animals
- Emojis (emotions)
- Places (Eiffel Tower, Pyramids etc.) 

**Inspired by WordPress.org support requests.**
https://wordpress.org/support/topic/icon-block/
https://wordpress.org/support/topic/icon-block-2/

### PR DESCRIPTION
**Pull request by EvanHerman**
Resolves #923 

Introduce the ability to use custom icons directly within the icon block.

No UI is introduced in this PR to manage custom icons (add/remove). Icons (`.svg` files) must be manually added to the proper directory (listed below) in your theme root. This PR lays the groundwork for the behind the scenes work - in a future release we could potentially create a UI to manage these icons.

### How To
1. Create a `coblocks/icons/` directory inside of your theme root. (eg: `wp-content/themes/primer/coblocks/icons/`)
2. Drop your `.svg` files into this directory
3. Check the icon block for your custom icons 🎉

When the `config.json` file (_see below_) is not present, these values are generated from the icons file name.
eg: `adobe-photoshop.svg` => Label: Adobe Photoshop, Keywords: `adobe, photoshop`

#### Optional Icon Configuration File
Users can also create an optional `config.json` file inside of the `coblocks/icons/` directory in the theme root. In this file, they can further define settings for each icon.
- `label` - The label for the icon
- `keywords` - The keywords used to search for the icon
- `icon_outlined` - The `name` of the outlined version of the icon (eg: `adobe-photoshop-outlined.svg`)

#### Demo Icons/config.json
[coblocks-custom-icons.zip](https://github.com/godaddy/coblocks/files/3730252/coblocks-custom-icons.zip)

### CODE REVIEW

**Comment by richtabor** [comment]
@paranoia1906 Do you mind updating this support ticket that we're working/experimenting on this: https://wordpress.org/support/topic/icon-block-2/

**Comment by AnthonyLedesma** [comment]
# How-To Use Custom Icons

The CoBlocks Icon block now allows the use of custom icons by examining the directory of your active theme in search of SVG files. This documentation will explain the steps and requirements needed to upload custom icons to use with CoBlocks.

## Steps

1. You should create a directory and subdirectory within the root of your active theme as follows: 
`coblocks/icons/` 
eg: `wp-content/themes/twentytwenty/coblocks/icons`.
2. Drop your `.svg` icons into this directory.
  * See [SVG Requirements](#svg-requirements)
3. Optionally configure the icons
  * See [Optional Configuration File](#optional-configuration-file)
4. Check the icon block for your custom icons 🎉

### SVG requirements
SVG files used for custom icons may not contain XML markup. For example, we can look at this sample `.svg` file.

**Original**
```html
<?xml version="1.0" encoding="iso-8859-1"?>
<svg height="20" viewBox="0 0 20 20" width="20" xmlns="http://www.w3.org/2000/svg"><circle class="inner-circle" cx="20" cy="20" r="8" stroke-width="8" stroke-dasharray="50.2655 50.2655" stroke-dashoffset="0"></circle></svg>
```

**Edited**
```html
<svg height="20" viewBox="0 0 20 20" width="20" xmlns="http://www.w3.org/2000/svg"><circle class="inner-circle" cx="20" cy="20" r="8" stroke-width="8" stroke-dasharray="50.2655 50.2655" stroke-dashoffset="0"></circle></svg>
```

### Optional Configuration File
Your custom icons can be further configured by placing a config.json file within the `coblocks/icons` folder alongside your SVG icons. The configuration file will allow you to specify search keywords, labels, and outline style.

Example `config.json` file:
```json
{
    "adobe-photoshop": {
        "label": "Adobe Photoshop",
        "keywords": "adobe photoshop",
        "icon_outlined": "adobe-photoshop-outlined.svg"
    },
    "apple-logo": {
        "label": "Apple Logo",
        "keywords": "apple logo",
        "icon_outlined": "apple-logo-outlined.svg"
    },
    "chrome": {
        "label": "Chrome",
        "keywords": "chrome web browser",
        "icon_outlined": "chrome-outlined.svg"
    },
    "mac-logo": {
        "label": "Mac Logo",
        "keywords": "mac logo computer",
        "icon_outlined": "mac-logo-outlined.svg"
    }
}
```

File: includes/class-coblocks-block-assets.php [hunk 1]
```diff
@@ -122,11 +122,87 @@ public function editor_assets() {
...
 
+	/**
+	 * Load custom icons from the theme directory, if they exist
+	 *
+	 * @return array Custom icons array if they exist, else empty array.
+	 */
+	public function get_custom_icons() {
+
+		$config = array();
+		$icons  = glob( get_stylesheet_directory() . '/coblocks/icons/*.svg' );
```
Comment (richtabor) [inline includes/class-coblocks-block-assets.php]: Does this work with child themes? Should this be `get_theme_file_uri` or `get_theme_file_path`? instead?

File: includes/class-coblocks-block-assets.php [hunk 2]
```diff
@@ -122,11 +122,87 @@ public function editor_assets() {
...
+
+				if ( ! array_key_exists( $icon, $custom_icons ) ) {
+
+					continue;
+
+				}
+
+				if ( array_key_exists( 'icon_outlined', $config[ $icon ] ) ) {
+
+					$metadata['icon_outlined'] = file_exists( get_stylesheet_directory() . '/coblocks/icons/' . $metadata['icon_outlined'] ) ? file_get_contents( get_stylesheet_directory() . '/coblocks/icons/' . $metadata['icon_outlined'] ) : '';
```
Comment (richtabor) [inline includes/class-coblocks-block-assets.php]: Same question, does this work with child themes?

File: includes/class-coblocks-block-assets.php (hunk 1 above)
Comment (AnthonyLedesma) [reply includes/class-coblocks-block-assets.php]: This does work with child themes. Specifically the custom icons must be uploaded to the child theme if you child theme is active.

File: includes/class-coblocks-block-assets.php (hunk 2 above)
Comment (AnthonyLedesma) [reply includes/class-coblocks-block-assets.php]: > This does work with child themes. Specifically the custom icons must be uploaded to the child theme if your child theme is active.

**Pull request merged**